### ISSUE TITLE
Consider returning new JWT on decode

### ISSUE THREAD

**Issue by rpcme** [jberger/Mojo-JWT#3]
I am having a difficult time managing JWTs when the decode method is stomping on values.  It also makes it difficult to test.

delete $self->{$_} for qw/claims expires not_before/;

Perhaps it may work to have the encode and decode methods be static.

**Comment by jberger** [comment]
that does sound intriguing, I'll give it a look soonish

**Comment by rpcme** [comment]
I guess the value of extending your module is when I can modify what ->claims produces.  Right now, I basically override the claims attribute so ->encode can take values based on some set of behaviours the programmer needs as per google's (or other) rules.

unfortunately, it seems as though ->decode does enough damage where I can't even decode and look at exp (which makes it so I can't validate the value).

another option is $jwt->plan (surely there's some name far more correct than this), which produces the claim that _would_ be encoded.

Most of this is talking out loud.

**Comment by jberger** [comment]
Yeah, I get your point. I was imagining that a user would create a new instance per message, in which case encode and decode make sense as they are, but that said, doing as you say would enforce that. I really do think it is worth consideration, I just haven't had a moment to look at it.

**Comment by rpcme** [comment]
Yep, no problem.  Maybe we can hash it out next month.

**Comment by strangelittlemonkey** [comment]
Is this still in the works?

**Comment by jberger** [comment]
The thought is still on my mind yes, I just haven't seen much need to actually do it. If you are interested in presenting a patch, I'd consider it.
